File: Practica-1-INF-121/POLIMORFISMO/EJERCICIO-9/Java/Minecraft.java
class BloqueCofre {
    private int capacidad;
    private int resistencia;
    private String tipo;

    public BloqueCofre(int capacidad, int resistencia, String tipo) {
        this.capacidad = capacidad;
        this.resistencia = resistencia;
        this.tipo = tipo;
    }

    public String accion() {
        return "Abriendo el cofre de tipo " + tipo + ".";
    }

    public String colocar(String orientacion) {
        return "Colocando el cofre en posición " + orientacion + ".";
    }

    public String romper() {
        return "Rompiendo el cofre de tipo " + tipo + ". Puede que se derramen los objetos dentro.";
    }
}

class BloqueTnt {
    private String tipo;
    private int daño;

    public BloqueTnt(String tipo, int daño) {
        this.tipo = tipo;
        this.daño = daño;
    }

    public String accion() {
        return "Activando el TNT de tipo " + tipo + ".";
    }

    public String colocar(String orientacion) {
        return "Colocando el TNT en posición " + orientacion + ".";
    }

    public String romper() {
        return "Rompiendo el TNT de tipo " + tipo + ". ¡Cuidado! Puede explotar.";
    }
}

class BloqueHorno {
    private String color;
    private int capacidadComida;

    public BloqueHorno(String color, int capacidadComida) {
        this.color = color;
        this.capacidadComida = capacidadComida;
    }

    public String accion() {
        return "Usando el horno de color " + color + ".";
    }

    public String colocar(String orientacion) {
        return "Colocando el horno en posición " + orientacion + ".";
    }

    public String romper() {
        return "Rompiendo el horno de color " + color + ". Puede que se rompa y se pierda comida.";
    }
}

public class Minecraft {
    public static void main(String[] args) {
        // Instanciar 2 bloques de cada tipo
        BloqueCofre cofre1 = new BloqueCofre(27, 5, "Madera");
        BloqueCofre cofre2 = new BloqueCofre(54, 10, "Metal");

        BloqueTnt tnt1 = new BloqueTnt("TNT Normal", 100);
        BloqueTnt tnt2 = new BloqueTnt("TNT Especial", 200);

        BloqueHorno horno1 = new BloqueHorno("Gris", 10);
        BloqueHorno horno2 = new BloqueHorno("Negro", 15);

        // Mostrar acciones y colocaciones
        System.out.println(cofre1.accion());
        System.out.println(cofre1.colocar("suelo"));
        System.out.println(cofre1.romper());
        System.out.println();

        System.out.println(cofre2.accion());
        System.out.println(cofre2.colocar("pared"));
        System.out.println(cofre2.romper());
        System.out.println();

        System.out.println(tnt1.accion());
        System.out.println(tnt1.colocar("suelo"));
        System.out.println(tnt1.romper());
        System.out.println();

        System.out.println(tnt2.accion());
        System.out.println(tnt2.colocar("pared"));
        System.out.println(tnt2.romper());
        System.out.println();

        System.out.println(horno1.accion());
        System.out.println(horno1.colocar("suelo"));
        System.out.println(horno1.romper());
        System.out.println();

        System.out.println(horno2.accion());
        System.out.println(horno2.colocar("pared"));
        System.out.println(horno2.romper());
    }
}

File: Practica-1-INF-121/POLIMORFISMO/EJERCICIO-9/Python/Minecraft.py
#Para los bloques del juego Minecraft se usará los siguientes objetos:
#La clase BloqueCofre con capacidad,resistencia y tipo. La clase BloqueTnt con tipo y daño. La clase BloqueHomo con color y capacidadComida 
#a) Crear e instanciar al menos 2 bloques de cada tipo 
#b) Sobrescribe accion() en BloqueCofre, BloqueTnt y BloqueHorno, mostrando distintos mensajes según el tipo de bloque.
#c) Sobrecarga colocar() para permitir colocar un bloque en diferentes orientaciones (por ejemplo, en el suelo o en la pared).
#d) Sobrescribe romper() en BloqueCofre, BloqueTnt y BloqueHorno, mostrando distintos mensajes según el tipo de bloque y que puede suceder al romperlos.

class BloqueCofre:
    def __init__(self, capacidad, resistencia, tipo):
        self.capacidad = capacidad
        self.resistencia = resistencia
        self.tipo = tipo

    def accion(self):
        return f"Abriendo el cofre de tipo {self.tipo}."

    def colocar(self, orientacion):
        return f"Colocando el cofre en posición {orientacion}."

    def romper(self):
        return f"Rompiendo el cofre de tipo {self.tipo}. Puede que se derramen los objetos dentro."


class BloqueTnt:
    def __init__(self, tipo, daño):
        self.tipo = tipo
        self.daño = daño

    def accion(self):
        return f"Activando el TNT de tipo {self.tipo}."

    def colocar(self, orientacion):
        return f"Colocando el TNT en posición {orientacion}."

    def romper(self):
        return f"Rompiendo el TNT de tipo {self.tipo}. ¡Cuidado! Puede explotar."


class BloqueHorno:
    def __init__(self, color, capacidad_comida):
        self.color = color
        self.capacidad_comida = capacidad_comida

    def accion(self):
        return f"Usando el horno de color {self.color}."

    def colocar(self, orientacion):
        return f"Colocando el horno en posición {orientacion}."

    def romper(self):
        return f"Rompiendo el horno de color {self.color}. Puede que se rompa y se pierda comida."


def main():
    # Instanciar 2 bloques de cada tipo
    cofre1 = BloqueCofre(capacidad=27, resistencia=5, tipo="Madera")
    cofre2 = BloqueCofre(capacidad=54, resistencia=10, tipo="Metal")

    tnt1 = BloqueTnt(tipo="TNT Normal", daño=100)
    tnt2 = BloqueTnt(tipo="TNT Especial", daño=200)

    horno1 = BloqueHorno(color="Gris", capacidad_comida=10)
    horno2 = BloqueHorno(color="Negro", capacidad_comida=15)

    # Mostrar acciones y colocaciones
    print(cofre1.accion())
    print(cofre1.colocar("suelo"))
    print(cofre1.romper())
    print()

    print(cofre2.accion())
    print(cofre2.colocar("pared"))
    print(cofre2.romper())
    print()

    print(tnt1.accion())
    print(tnt1.colocar("suelo"))
    print(tnt1.romper())
    print()

    print(tnt2.accion())
    print(tnt2.colocar("pared"))
    print(tnt2.romper())
    print()

    print(horno1.accion())
    print(horno1.colocar("suelo"))
    print(horno1.romper())
    print()

    print(horno2.accion())
    print(horno2.colocar("pared"))
    print(horno2.romper())


if __name__ == "__main__":
    main()